src/diffusers/loaders/ip_adapter_utils.py

```diff
@@ -0,0 +1,144 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import copy
+from typing import TYPE_CHECKING, Dict, List, Union
+
+from ..utils import logging
+
+
+if TYPE_CHECKING:
+    # import here to avoid circular imports
+    from ..models import UNet2DConditionModel
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+def _translate_into_actual_layer_name(name):
+    """Translate user-friendly name (e.g. 'mid') into actual layer name (e.g. 'mid_block.attentions.0')"""
+    if name == "mid":
+        return "mid_block.attentions.0"
+
+    updown, block, attn = name.split(".")
+
+    updown = updown.replace("down", "down_blocks").replace("up", "up_blocks")
+    block = block.replace("block_", "")
+    attn = "attentions." + attn
+
+    return ".".join((updown, block, attn))
+
+
+def _maybe_expand_ip_scales(unet: "UNet2DConditionModel", weight_scale: List[Union[float, Dict]]):
+    blocks_with_transformer = {
+        "down": [i for i, block in enumerate(unet.down_blocks) if hasattr(block, "attentions")],
+        "up": [i for i, block in enumerate(unet.up_blocks) if hasattr(block, "attentions")],
+    }
+    transformer_per_block = {"down": unet.config.layers_per_block, "up": unet.config.layers_per_block + 1}
+
+    expanded_weight_scales = _maybe_expand_scales_for_one_ip_adapter(
+                                weight_scale, blocks_with_transformer, transformer_per_block, unet.state_dict()
+                            )
+
+    return expanded_weight_scales
+
+
+def _maybe_expand_scales_for_one_ip_adapter(
```

Comment:
> why no use the same as the LoRAs?
> 
> https://github.com/huggingface/diffusers/blob/cf6e0407e051467b480830d3ed97d2873b5019d3/src/diffusers/loaders/unet_loader_utils.py#L27
> 
> or maybe use a `copied from`? @yiyixuxu what would be the best solution here?

Actually we want something slightly different from ```_maybe_expand_lora_scales_for_one_adapter```, specifically:
https://github.com/huggingface/diffusers/blob/1f6b84a5b33c741efd48f5493f50a2996ca38fc7/src/diffusers/loaders/ip_adapter_utils.py#L121-L128
where we want a default scale=0 and a check to avoid assigning scales that does not match the number of blocks. But yes, ```_translate_into_actual_layer_name``` and ```_maybe_expand_ip_scales``` are the same as the LoRAs. I'm not sure about the best solution. Should I try to modify the LoRAs or just keep the new ```ip_adapter_utils```?

Reply:
Both have the same time complexity, but IMO I like it more with your implementation, so if we don't set the scale for a block, we can assume we want it to be 0 and also has more checks for errors.

I'm in favor of modifying the original LoRA function to just keep one function for this and future uses but let's wait for @yiyixuxu opinion.

cc'ing also @UmerHA for insights in case I missed a reason why it's important to keep the scale at 1

if we want to keep the same value for the LoRAs, maybe we can add a `default_scale` arg

Reply:
Yeah I agree. In our case we prefer a default scale=0 to keep our ```scale_config``` succinct, but I think maybe a default scale=1 is more suitable to LoRAs.

Reply:
When I was testing the blockwise lora too for this PR, I found that I had to manually set to 0 all of them because I was testing how each block reflected in the result.

IMO when someone use this granular kind of control expects  them to default to 0 instead of 1, but at the end is just a matter of preference and I agree that is better to not change the existing behavior. 

BTW, I have done some experiments with your PR, I'll post them soon so we can discuss them and I'll appreciate your help if you're interested.